### PR TITLE
libkb: chain link parsing V2

### PR DESCRIPTION
**Pull request by maxtaco**
- infer all fields aside form those signed by user (or stubbed summaries)
- BW savings and simpler code

### CODE REVIEW

**Comment by maxtaco** [comment]
image is mainly due to simplifying the sig/get structure in the test vectors.

File: go/kbcrypto/nacl_sig_info.go [hunk 1]
```diff
@@ -195,3 +197,12 @@ func NaclVerifyWithPayload(sig string, payloadIn []byte) (nk *NaclSigningKeyPubl
 
 	return nk, fullBody, nil
 }
+
+func (s NaclSigInfo) SigID() (ret keybase1.SigID, err error) {
+	var body []byte
+	body, err = EncodePacketToBytes(&s)
+	if err != nil {
+		return ret, nil
```
Comment (zapu) [inline go/kbcrypto/nacl_sig_info.go]: `ret, nil`?

Comment (maxtaco) [reply go/kbcrypto/nacl_sig_info.go]: Thanks for finding this!!! Great eye!

File: go/libkb/chain_link_parse_2.go [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+func newSigChainPayloadJSON(s string) *sigChainPayloadJSON {
+	return &sigChainPayloadJSON{s: s, b: []byte(s)}
+}
+
+func newSigChainPayloadJSONFromBytes(b []byte) *sigChainPayloadJSON {
+	return &sigChainPayloadJSON{s: string(b), b: b}
+}
+
+func (s *sigChainPayloadJSON) String() string {
+	return s.s
```
Comment (zapu) [inline go/libkb/chain_link_parse_2.go]: I don't see `s string` or `s.String` being used. I even removed both to try to compile and it still did. Will it be used in the future?

**Comment by maxtaco** [comment]
thanks for the review!

File: go/libkb/chain_link_parse_2.go [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+func (s *sigChainPayloadJSON) Hash() LinkID {
+	return ComputeLinkID(s.b)
+}
+
+func getPayloadJSONFromServerLink(packed []byte) (*sigChainPayloadJSON, error) {
+	data, _, _, err := jsonparserw.Get(packed, "payload_json")
+	if err != nil {
+		return nil, err
+	}
+	sdata, err := strconv.Unquote(`"` + string(data) + `"`)
```
Comment (heronhaye) [inline go/libkb/chain_link_parse_2.go]: as long as we're imploding stuff, does it make sense to deserialize payload_json on the server to save on all this escaping? or would it be a perf hit?

File: go/libkb/chain_link_parse_2.go [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+	err = importServerTrustFields(m, &tmp, data, selfUID)
+	if err != nil {
+		return nil, err
+	}
+
+	ret.unpacked = &tmp
+
+	// this should be safe, but please double-check in review
+	ret.hashVerified = true
+	ret.payloadVerified = true
```
Comment (heronhaye) [inline go/libkb/chain_link_parse_2.go]: i checked this, seemed to be OK, checking it on importing.

File: go/libkb/chain_link_parse_2.go [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+	if err != nil || versionRaw == 0 {
+		return nil, ChainLinkError{"cannot read signature version from server"}
+	}
+
+	pgpSig := getPGPSig(m, data)
+	sigVersion := SigVersion(versionRaw)
+
+	var ir *importRes
+	switch {
+	case sigVersion == KeybaseSignatureV1 && pgpSig != "":
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/libkb/chain_link_parse_2.go]: it is server trust. We assert in this file that V1s are JSON And V2s are msgpack, to handle the polyglots as you say.

File: go/libkb/chain_link_parse_2.go [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+		seqno:               ol.Seqno,
+		seqType:             ol.SeqType,
+		ignoreIfUnsupported: ol.IgnoreIfUnsupported,
+		highSkip:            highSkipPtr,
+		sigVersion:          ol.Version,
+		outerLinkV2:         ol,
+		stubbed:             true,
+	}
+	ret = NewChainLink(m.G(), parent, linkID)
+	ret.unpacked = unpacked
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/libkb/chain_link_parse_2.go]: makes sense

File: go/libkb/chain_link_parse_2.go (hunk 3 above)
Comment (maxtaco) [reply go/libkb/chain_link_parse_2.go]: hmmm, then we'll have to worry about the canonicalization of the JSON, but it should be canonicalized....

**Pull request merged**